src/Client/HedgedConnections.cpp

```diff
@@ -0,0 +1,416 @@
+
+    for (auto & other_replica : replicas[replica->parallel_replica_offset])
+    {
+        if (other_replica->isReady() && other_replica != replica)
+        {
+            other_replica->connection->sendCancel();
+            finishProcessReplica(other_replica, true);
+        }
+    }
+
+    /// If we received data from replicas with all offsets, we need to stop choosing new replicas.
+    if (get_hedged_connections.hasEventsInProcess() && offsets_with_received_data.size() == replicas.size())
+    {
+        get_hedged_connections.stopChoosingReplicas();
+        if (next_replica_in_process)
+            epoll.remove(get_hedged_connections.getFileDescriptor());
+    }
+}
+
+void HedgedConnections::processTimeoutEvent(ReplicaStatePtr & replica, TimerDescriptorPtr timeout_descriptor)
+{
+    epoll.remove(timeout_descriptor->getDescriptor());
+    replica->active_timeouts.erase(timeout_descriptor->getDescriptor());
+    timeout_fd_to_replica.erase(timeout_descriptor->getDescriptor());
+
+    if (timeout_descriptor->getType() == TimerTypes::RECEIVE_TIMEOUT)
+    {
+        size_t offset = replica->parallel_replica_offset;
+        finishProcessReplica(replica, true);
+
+        /// Check if there is no active connections with the same offset.
+        if (active_connections_count_by_offset[offset] == 0)
+            throw NetException("Receive timeout expired", ErrorCodes::SOCKET_TIMEOUT);
+    }
+    else if (timeout_descriptor->getType() == TimerTypes::RECEIVE_DATA_TIMEOUT)
+    {
+        offsets_queue.push(replica->parallel_replica_offset);
+        tryGetNewReplica();
+    }
+}
+
+void HedgedConnections::tryGetNewReplica()
+{
+    ReplicaStatePtr new_replica = get_hedged_connections.getNextConnection(/*non_blocking*/ true);
+
+    /// Skip replicas that doesn't support two-level aggregation if we didn't disable it in sendQuery.
+    while (new_replica->isReady() && !disable_two_level_aggregation
+           && new_replica->connection->getServerRevision(get_hedged_connections.getConnectionTimeouts()) < DBMS_MIN_REVISION_WITH_CURRENT_AGGREGATION_VARIANT_SELECTION_METHOD)
+        new_replica = get_hedged_connections.getNextConnection(/*non_blocking*/ true);
+
+    if (new_replica->isReady())
+    {
+        new_replica->parallel_replica_offset = offsets_queue.front();
+        offsets_queue.pop();
+        replicas[new_replica->parallel_replica_offset].push_back(new_replica);
+        epoll.add(new_replica->fd);
+        fd_to_replica[new_replica->fd] = new_replica;
+        ++active_connections_count_by_offset[new_replica->parallel_replica_offset];
+        pipeline_for_new_replicas.run(new_replica);
+    }
+    else if (new_replica->isNotReady() && !next_replica_in_process)
+    {
+        epoll.add(get_hedged_connections.getFileDescriptor());
+        next_replica_in_process = true;
+    }
+
+    if (next_replica_in_process && (new_replica->isCannotChoose() || offsets_queue.empty()))
```

Comment:
So, what if `new_replica->isCannotChoose()` and we still need a replica? Shouldn't we throw exception?